# 01 - Dataset Understanding

Primer notebook para entender la estructura del dataset del challenge de Analytics Engineer.

Objetivos:
- Cargar el CSV.
- Revisar dimensiones, columnas y tipos de datos.
- Entender nulos, duplicados y cardinalidades.
- Validar la granularidad de la tabla.
- Identificar primeras hipotesis para el analisis de sellers.

## 1. Imports

In [1]:
import pandas as pd

from data_loading import (
    load_dataset,
    print_dataset_report,
    get_dataset_overview,
    get_column_profile,
    get_numeric_profile,
    get_categorical_profile,
    get_seller_grain_checks,
)
print("Imports OK")


Imports OK


## 2. Carga del dataset

In [2]:
DATA_PATH = "df_challenge_meli.csv"

df = load_dataset(DATA_PATH)
df.head(10)

,tim_day,seller_nickname,titulo,seller_reputation,stock,logistic_type,condition,is_refurbished,price,regular_price,categoria,url,category_id,category_name
0,2024-08-01,5a39342818,Gel Cerave Limpiador Espumoso Piel Normal A Gr...,green_platinum,16155,FBM,new,False,260.93,355.00,BEAUTY PERSONAL CARE,https://articulo.mercadolibre.com.mx/MLM-17883...,BEAUTY PERSONAL CARE,CUIDADO PERSONAL Y BELLEZA
1,2024-08-01,5a39342818,Kit Garnier Express Aclara Serum +crema +gel C...,green_platinum,6253,FBM,new,False,344.26,NaN,BEAUTY PERSONAL CARE,https://articulo.mercadolibre.com.mx/MLM-31259...,BEAUTY PERSONAL CARE,CUIDADO PERSONAL Y BELLEZA
2,2024-08-01,66d9659692,3 Piezas De Minoxidil 5% Anacastel,green_platinum,10066,FBM,new,False,363.00,660.00,HAIR,https://articulo.mercadolibre.com.mx/MLM-19360...,HAIR,CUIDADO DEL CABELLO
3,2024-08-01,5a39342818,Limpiador Cerave Control Imperfecciones Para P...,green_platinum,19678,FBM,new,False,241.00,410.00,BEAUTY PERSONAL CARE,https://articulo.mercadolibre.com.mx/MLM-18485...,BEAUTY PERSONAL CARE,CUIDADO PERSONAL Y BELLEZA
4,2024-08-01,5a39342818,Labial Líquido Maybelline Super Stay Matte Ink...,green_platinum,4862,FBM,new,False,133.47,152.49,FRAGRANCES & MAKE UP,https://articulo.mercadolibre.com.mx/MLM-15825...,FRAGRANCES & MAKE UP,FRAGANCIAS Y MAQUILLAJE
5,2024-08-01,5a39342818,Serúm Garnier Express Booster 4% 30ml Momento ...,green_platinum,30050,FBM,new,False,135.78,204.00,BEAUTY PERSONAL CARE,https://articulo.mercadolibre.com.mx/MLM-28760...,BEAUTY PERSONAL CARE,CUIDADO PERSONAL Y BELLEZA
6,2024-08-01,5a39342818,Kit Garnier Rutina Express Aclara Anti Acné Ác...,green_platinum,3401,FBM,new,False,541.00,NaN,BEAUTY PERSONAL CARE,https://articulo.mercadolibre.com.mx/MLM-22521...,BEAUTY PERSONAL CARE,CUIDADO PERSONAL Y BELLEZA
7,2024-08-01,5a39342818,Garnier Express Aclara Serum Anti Manchas Con ...,green_platinum,14161,FBM,new,False,142.94,215.00,BEAUTY PERSONAL CARE,https://articulo.mercadolibre.com.mx/MLM-28760...,BEAUTY PERSONAL CARE,CUIDADO PERSONAL Y BELLEZA
8,2024-08-01,5a39342818,L'oréal Paris Uv Defender Fluido Invisible Pro...,green_platinum,10685,FBM,new,False,208.50,NaN,BEAUTY PERSONAL CARE,https://articulo.mercadolibre.com.mx/MLM-14773...,BEAUTY PERSONAL CARE,CUIDADO PERSONAL Y BELLEZA
9,2024-08-01,5a39342818,Sérum Anti-manchas Glycolic Bright Con Ácido G...,green_platinum,6627,FBM,new,False,380.00,NaN,BEAUTY PERSONAL CARE,https://articulo.mercadolibre.com.mx/MLM-30787...,BEAUTY PERSONAL CARE,CUIDADO PERSONAL Y BELLEZA


## 3. Overview general

In [3]:
overview = get_dataset_overview(df)
overview

{'rows': 185250,
 'columns': 14,
 'memory_mb': np.float64(140.14),
 'duplicated_rows': 0,
 'min_tim_day': Timestamp('2024-08-01 00:00:00'),
 'max_tim_day': Timestamp('2024-08-01 00:00:00'),
 'distinct_tim_days': 1,
 'distinct_sellers': 46586,
 'distinct_urls': 185250,
 'distinct_category_ids': 54,
 'distinct_category_names': 54}

## 4. Perfil de columnas

In [4]:
column_profile = get_column_profile(df)
column_profile

,column,dtype,null_count,null_pct,unique_count,unique_pct,sample_values
0,tim_day,datetime64[ns],0,0.00,1,0.00,[2024-08-01]
1,seller_nickname,object,0,0.00,46586,25.15,"[5a39342818, 66d9659692, 3059e1c98b, 92a7303a2..."
2,titulo,object,0,0.00,174746,94.33,[Gel Cerave Limpiador Espumoso Piel Normal A G...
3,seller_reputation,object,2372,1.28,9,0.00,"[green_platinum, green_gold, green, green_silv..."
4,stock,int64,0,0.00,3609,1.95,"[16155, 6253, 10066, 19678, 4862]"
5,logistic_type,object,0,0.00,5,0.00,"[FBM, XD, DS, Otro, FLEX]"
6,condition,object,0,0.00,3,0.00,"[new, used, not_specified]"
7,is_refurbished,bool,0,0.00,2,0.00,"[False, True]"
8,price,float64,1516,0.82,36529,19.72,"[260.93, 344.26, 363.0, 241.0, 133.47]"
9,regular_price,float64,135294,73.03,12109,6.54,"[355.0, 660.0, 410.0, 152.49, 204.0]"


## 5. Nulos ordenados por impacto

In [5]:
column_profile.sort_values("null_pct", ascending=False)[
    ["column", "dtype", "null_count", "null_pct", "unique_count"]
]

,column,dtype,null_count,null_pct,unique_count
9,regular_price,float64,135294,73.03,12109
3,seller_reputation,object,2372,1.28,9
8,price,float64,1516,0.82,36529
0,tim_day,datetime64[ns],0,0.00,1
1,seller_nickname,object,0,0.00,46586
2,titulo,object,0,0.00,174746
4,stock,int64,0,0.00,3609
5,logistic_type,object,0,0.00,5
6,condition,object,0,0.00,3
7,is_refurbished,bool,0,0.00,2


## 6. Estadisticas numericas

In [6]:
numeric_profile = get_numeric_profile(df)
numeric_profile

,column,count,mean,std,min,1%,5%,25%,50%,75%,95%,99%,max
0,stock,185250.0,307.017090,3.367217e+03,0.0,0.0,1.0,2.0000,8.0,42.00,583.550,4999.51,9.999900e+04
1,price,183734.0,37015.311261,1.124158e+07,0.0,55.0,98.0,270.0000,568.1,1387.00,7856.624,29999.00,4.772354e+09
2,regular_price,49956.0,1614.429170,8.408646e+03,6.1,69.0,120.0,278.6725,499.0,1099.99,5600.000,19473.00,9.999990e+05


## 7. Valores frecuentes en columnas categoricas

In [7]:
categorical_profile = get_categorical_profile(df, top_n=10)

for column, profile in categorical_profile.items():
    print(f"\n{column}")
    display(profile)


seller_nickname


,seller_nickname,records,pct
0,a9847ce81d,1421,0.77
1,5a39342818,877,0.47
2,77b179c68c,876,0.47
3,d5f5bc5f45,795,0.43
4,02cdf9f94d,510,0.28
5,a6d42f133e,484,0.26
6,92d2275dde,462,0.25
7,55c5031bc1,429,0.23
8,ca47ea1279,388,0.21
9,73479d561b,384,0.21



titulo


,titulo,records,pct
0,Memoria Usb Adata Uv250 Auv250-16g-rbk 16gb 2....,100,0.05
1,Apple AirPods Pro 2nda Generación Color De La ...,80,0.04
2,Tarjeta De Memoria Kingston Sdcs2 Canvas Sele...,76,0.04
3,Ariana Grande Ari Edp 100 Ml,73,0.04
4,Apple Cargador 20w Usb-c Lightning To Usb Colo...,54,0.03
5,Impresora Multifunción A Color Hp Deskjet Ink ...,53,0.03
6,Perry Ellis 360º Red Eau De Toilette 100 ml Pa...,51,0.03
7,Interfaz De Audio Behringer U-phoria Umc202hd,51,0.03
8,Marvel's Sentinel - Figura De Acción Titan Her...,42,0.02
9,Pantalla Hisense 32h5g 32'' Hd Dts Virtual 202...,40,0.02



seller_reputation


,seller_reputation,records,pct
0,green_platinum,69625,37.58
1,green,33838,18.27
2,green_gold,28332,15.29
3,green_silver,24731,13.35
4,newbie,9221,4.98
5,yellow,7245,3.91
6,red,4219,2.28
7,light_green,3101,1.67
8,orange,2566,1.39
9,<NULL>,2372,1.28



logistic_type


,logistic_type,records,pct
0,XD,116763,63.03
1,FBM,31483,16.99
2,DS,24645,13.30
3,Otro,10541,5.69
4,FLEX,1818,0.98



condition


,condition,records,pct
0,new,169616,91.56
1,used,15551,8.39
2,not_specified,83,0.04



is_refurbished


,is_refurbished,records,pct
0,False,184515,99.6
1,True,735,0.4



categoria


,categoria,records,pct
0,SPORTS,14252,7.69
1,HEALTH,12587,6.79
2,"BOOKS, MULTIMEDIA & OTHER E!",10355,5.59
3,TOYS AND GAMES,10251,5.53
4,ACC CARS & VANS,10128,5.47
5,OTHER,7797,4.21
6,HOUSEHOLD APPLIANCES,7090,3.83
7,ANTIQUES & HOBBIES,6453,3.48
8,HOUSEHOLD ITEMS,5569,3.01
9,APPAREL ACCESSORIES,5423,2.93



url


,url,records,pct
0,https://articulo.mercadolibre.com.mx/MLM-17883...,1,0.0
1,https://articulo.mercadolibre.com.mx/MLM-93384...,1,0.0
2,https://articulo.mercadolibre.com.mx/MLM-19328...,1,0.0
3,https://articulo.mercadolibre.com.mx/MLM-15076...,1,0.0
4,https://articulo.mercadolibre.com.mx/MLM-61188...,1,0.0
5,https://articulo.mercadolibre.com.mx/MLM-24655...,1,0.0
6,https://articulo.mercadolibre.com.mx/MLM-15379...,1,0.0
7,https://articulo.mercadolibre.com.mx/MLM-13503...,1,0.0
8,https://articulo.mercadolibre.com.mx/MLM-83275...,1,0.0
9,https://articulo.mercadolibre.com.mx/MLM-14162...,1,0.0



category_id


,category_id,records,pct
0,SPORTS,14252,7.69
1,HEALTH,12587,6.79
2,"BOOKS, MULTIMEDIA & OTHER E!",10355,5.59
3,TOYS AND GAMES,10251,5.53
4,ACC CARS & VANS,10128,5.47
5,OTHER,7797,4.21
6,HOUSEHOLD APPLIANCES,7090,3.83
7,ANTIQUES & HOBBIES,6453,3.48
8,HOUSEHOLD ITEMS,5569,3.01
9,APPAREL ACCESSORIES,5423,2.93



category_name


,category_name,records,pct
0,DEPORTES,14252,7.69
1,SALUD,12587,6.79
2,"LIBROS, MULTIMEDIA Y OTROS",10355,5.59
3,JUGUETES Y JUEGOS,10251,5.53
4,ACCESORIOS PARA AUTOS Y CAMIONETAS,10128,5.47
5,OTROS,7797,4.21
6,ELECTRODOMÉSTICOS,7090,3.83
7,ANTIGÜEDADES Y PASATIEMPOS,6453,3.48
8,ARTÍCULOS DEL HOGAR,5569,3.01
9,ACCESORIOS DE VESTIR,5423,2.93


## 8. Checks de granularidad

El caso pide clusterizar sellers, pero el dataset parece estar a nivel item/publicacion. Estos checks ayudan a validarlo.

In [8]:
grain_checks = get_seller_grain_checks(df)
grain_checks

{'avg_rows_per_seller': 3.98,
 'median_rows_per_seller': 1.0,
 'max_rows_per_seller': 1421,
 'sellers_with_one_row': 24517,
 'avg_unique_urls_per_seller': 3.98,
 'median_unique_urls_per_seller': 1.0,
 'max_unique_urls_per_seller': 1421,
 'avg_days_per_seller': 1.0,
 'max_days_per_seller': 1}

In [9]:
df.groupby("seller_nickname").size().describe()

count    46586.000000
mean         3.976517
std         14.468409
min          1.000000
25%          1.000000
50%          1.000000
75%          3.000000
max       1421.000000
dtype: float64

## 9. Primeras vistas por seller

In [10]:
seller_preview = (
    df.groupby("seller_nickname")
    .agg(
        rows=("seller_nickname", "size"),
        unique_urls=("url", "nunique"),
        categories=("category_name", "nunique"),
        avg_price=("price", "mean"),
        median_price=("price", "median"),
        total_stock=("stock", "sum"),
    )
    .sort_values("rows", ascending=False)
)

seller_preview.head(20)

,rows,unique_urls,categories,avg_price,median_price,total_stock
seller_nickname,,,,,,
a9847ce81d,1421,1421,48,3543.972969,1652.000,55364
5a39342818,877,877,12,322.262611,189.000,658801
77b179c68c,876,876,32,1513.220674,903.315,81954
d5f5bc5f45,795,795,1,904.283019,700.000,7873
02cdf9f94d,510,510,11,2078.318294,1217.000,276217
a6d42f133e,484,484,2,882.822539,740.000,3326
92d2275dde,462,462,7,522.126818,429.000,48335
55c5031bc1,429,429,17,997.367809,449.210,14574
ca47ea1279,388,388,40,3477.487835,820.870,3829


## 10. Reporte completo

Esta celda imprime todo el reporte en formato texto. Puede ser util para una revision rapida.

In [11]:
report = print_dataset_report(df)


Resumen general del dataset
rows: 185250
columns: 14
memory_mb: 145.11
duplicated_rows: 0
min_tim_day: 2024-08-01 00:00:00
max_tim_day: 2024-08-01 00:00:00
distinct_tim_days: 1
distinct_sellers: 46586
distinct_urls: 185250
distinct_category_ids: 54
distinct_category_names: 54

Perfil de columnas
           column          dtype  null_count  null_pct  unique_count  unique_pct                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         sampl

## 11. Notas y próximos pasos

Hallazgos de esta exploración inicial:

- **Granularidad:** el dataset está a nivel ítem/publicación (≈4 ítems por seller en promedio); para segmentar vendedores hay que consolidar el grano a nivel seller.
- **Calidad:** destaca un precio máximo de ~$4,77 mil millones MXN (error de carga), `regular_price` nulo en el 73% (comportamiento esperado) y `price` nulo en 0,82%.
- **Logística:** 5 canales (XD 63%, FBM 17%, DS 13,3%, Otro 5,7%, FLEX <1%) → `pct_ds` es relevante.
- **Features candidatas:** precio (log), volumen (ítems/stock), proporciones logísticas, reputación (con flag `has_reputation`) y descuentos.

El flujo completo (limpieza, feature engineering, las 3 iteraciones del clustering, los 5 segmentos y la estrategia) está en **[`02_full_challenge_flow.ipynb`](02_full_challenge_flow.ipynb)** y en **[`INFORME_TECNICO.md`](INFORME_TECNICO.md)**.